File: langs/haskell/prjs/isses/M001.hs
import Data.List (transpose,tails,isPrefixOf,permutations)
import qualified Data.Map as Map
import Data.Maybe (fromMaybe, fromJust)
import Control.Monad(guard, MonadPlus, mzero, mplus, when)
import Control.Monad.ST (ST)
import Control.Applicative(pure)
import Numeric(showHex, readInt)
import Data.Char(digitToInt)

import FromFile(user_f)
import Tester(zz_act,pad_n)
import FromString(toI,to_nums,neg1)
import Util(isin,notin,show_nss)

m001 :: [String] -> [Int]
m001 (s:_) =  to_nums s
test_m = let res = m001 ["10 3 2 5 7 8"] in show res  -- exp is

setter :: a -> Int -> [a] -> [a]
setter v i xs = f ++ (v:r) where f = take i xs; r = drop (i+1) xs
test_s = let res = setter 10 0 [0,0,0,0,0] in show res

domain :: (Num a, Ord a) => [a] -> [a] -> a
domain ns dp = max (maximum (foldl(folder1) dp [0..(length dp)-1])) (maximum (foldl(folder2) dp [0..(length dp)-1])) where
  folder1 ac i = let ac1 = setter (ns!!i) i ac in if i > 0 
    then let m = max (ac1!!i) (ac1!!(i-1)) in let ac2 = setter m i ac1 in if i > 1
      then let m2 = max (ac2!!i) (ac2!!(i-2) + ns!!i) in setter m2 i ac2
      else ac2
    else ac1
  folder2 ac i = let ac1 = setter (ns!!(i+1)) i  ac in if i > 0
    then let m = max (ac1!!i) (ac1!!(i-1)) in let ac2 = setter m i ac1 in if i > 1
      then let m2 = max (ac2!!i) (ac2!!(i-2) + ns!!(i+1)) in setter m2 i ac2
      else ac2
    else ac1
test = let res = domain [10,3,2,5,7,8] [0,0,0,0,0] in show res

domains :: [String] -> [String]
domains ss = let ns = m001 ss; dp = replicate (length ns - 1) 0 in [show $domain ns dp]
test_ds = let res = domains ["10 3 2 5 7 8"] in show res

user :: IO [String]; user = do 
  -- line <- getLine; o
  -- let lim  = (length line) - 2
  -- tbls <- sequence [getLine | _ <- [0..1]]
  line1 <- getLine
  line2 <- getLine
  return ([line1,line2])

paiza :: IO String; 
paiza = do
  arg <- user
  -- return (show arg)
  let ans = domains arg in return (unlines ans)

from_file :: FilePath -> IO [String]; from_file path = user_f path

developer:: String -> String -> String -> IO String; developer issue result name =  do
  arg <- from_file issue; exp <- from_file result; let ans = domains(arg)
  return(zz_act "M001 Test" ans exp name)


develop :: IO String; develop = do
  let issue1 = "a:/pj/mamo/refs/isses/issue1.txt"; result1 = "a:/pj/mamo/refs/isses/result1.txt"
  let issue2 = "a:/pj/mamo/refs/isses/issue2.txt"; result2 = "a:/pj/mamo/refs/isses/result2.txt"
  let issue3 = "a:/pj/mamo/refs/isses/issue3.txt"; result3 = "a:/pj/mamo/refs/isses/result3.txt"
  let issue4 = "a:/pj/mamo/refs/isses/issue4.txt"; result4 = "a:/pj/mamo/refs/isses/result4.txt"
  let issue5 = "a:/pj/mamo/refs/isses/issue5.txt"; result5 = "a:/pj/mamo/refs/isses/result5.txt"
  let issue6 = "a:/pj/mamo/refs/isses/issue6.txt"; result6 = "a:/pj/mamo/refs/isses/result6.txt"
  res1 <- developer issue1 result1 "test 1"; res2 <- developer issue2 result2 "test 2"
  res3 <- developer issue3 result3 "test 3"; res4 <- developer issue4 result4 "test 4"
  res5 <- developer issue5 result5 "test 5"; res6 <- developer issue6 result6 "test 6"
  -- return (res1 ++ "\n" ++ res2 ++ "\n" ++ res3 ++ "\n" ++ res4)
  -- return (res1 ++ "\n" ++ res2 ++ "\n" ++ res3 ++ "\n" ++ res4 ++ "\n" ++ res5)
  return (res1 ++ "\n" ++ res2 ++ "\n" ++ res3 ++ "\n" ++ res4 ++ "\n" ++ res5 ++ "\n" ++ res6)

refactor :: IO String; refactor = do
  let issue  = "a:/pj/mamo/refs/isses/issue1.txt"; result = "a:/pj/mamo/refs/isses/result1.txt"
  arg  <- from_file  issue; exp <- from_file result
  -- return((unlines arg) ++ (unlines exp))
  return (zz_act "M001 Test" (domains(arg)) exp  "refactor")

-- M001 m001 domain domains user permutations
main = do
  -- ans <- paiza
  ans <- develop
  -- ans <- refactor
  -- let ans = test
  putStrLn ans

File: langs/java/prjs/isses/M001.java
import java.io.InputStreamReader;
import java.io.BufferedReader;
import java.io.IOException;
import java.util.List;
import java.util.ArrayList;
import java.util.Set;
import java.util.Map;
import java.util.HashSet;
import java.util.HashMap;
import java.util.Optional;
import java.util.stream.Stream;
import java.util.stream.IntStream;
import java.util.stream.Collectors;
import java.util.Arrays;
import java.util.Queue;
import java.util.LinkedList;
import static java.util.Arrays.asList;
import static java.util.stream.Collectors.toList;
import static java.util.stream.Collectors.joining;
import static java.util.stream.Collectors.toCollection;
import static java.util.Collections.reverse;
import static java.util.Collections.max;
import static java.util.Collections.min;

public class M001{
  public static void main(String[] args) throws Exception{
    // refactor();
    develop();
    // product();
  } // m001 M001 domain input
  
  public static void refactor() throws Exception {
    final String issue = "a:/pj/mamo/refs/isses/issue1.txt"; final String result = "a:/pj/mamo/refs/isses/result1.txt";
    final UserFile uf = new UserFile(); uf.ffUser(result); final List<String> exp = uf.list(); uf.ffUser(issue);
    // System.out.println(uf.list()); System.out.println(exp);
    M001 m001 = new M001(uf); m001.test();
    // developer(issue,result,"refactor");
  }
  public static void product() throws Exception{ 
    User user = new User(); 
    // System.out.println(user.list());
    M001 m001 = new M001(user);
    List<String> ans = m001.domains(); System.out.println(ans.stream().collect(joining("\n")));
  }
  public static void develop () throws Exception {
    final String issue1 = "a:/pj/mamo/refs/isses/issue1.txt"; final String result1 = "a:/pj/mamo/refs/isses/result1.txt";
    final String issue2 = "a:/pj/mamo/refs/isses/issue2.txt"; final String result2 = "a:/pj/mamo/refs/isses/result2.txt";
    final String issue3 = "a:/pj/mamo/refs/isses/issue3.txt"; final String result3 = "a:/pj/mamo/refs/isses/result3.txt";
    final String issue4 = "a:/pj/mamo/refs/isses/issue4.txt"; final String result4 = "a:/pj/mamo/refs/isses/result4.txt";
    final String issue5 = "a:/pj/mamo/refs/isses/issue5.txt"; final String result5 = "a:/pj/mamo/refs/isses/result5.txt";
    final String issue6 = "a:/pj/mamo/refs/isses/issue6.txt"; final String result6 = "a:/pj/mamo/refs/isses/result6.txt";
    developer(issue1, result1, "test-1"); developer(issue2, result2, "test-2");
    developer(issue3, result3, "test-3"); developer(issue4, result4, "test-4");
    developer(issue5, result5, "test-5"); developer(issue6, result6, "test-6");
  } // main
  public static void developer(String issue, String result, String name) throws Exception{
    UserFile uf = new UserFile(); uf.ffUser(result); List<String> exp = uf.list(); uf.ffUser(issue);
    M001 m001 = new M001(uf); List<String> ans =  m001.domains(); Tester<String> t = new Tester<>("M001 Test");
    t.test(name, ans.toString(),exp.toString());
  }
  public final List<String> lines; public final FromString fs;
  private List<String> all = new ArrayList<>();

  public M001(UserFile user){ this.lines = user.list(); this.fs = new FromString();}
  
  private void m001() {
    String[] fst = this.lines.get(0).split(" "), snd = this.lines.get(1).split(" ");
    this.all.addAll(asList(fst)); this.all.addAll(asList(snd));
  }
  // public void test(){ m001(); System.out.println(this.all);} // exp is 

  private Map<String,Integer> domain(List<String> ss) {
    Map<String,Integer> res = new HashMap<>();
    for(String s: ss){ res.put(s,res.get(s) == null ? 1 : res.get(s) + 1);}
    return res;
  }
  public void test_d(){
    List<String> arg = asList("fishing","gardening","swimming","fishing","hunting","fishing","fishing","biting");
    Map<String,Integer> res = domain(arg); System.out.println(res);
  }  // exp is exp

  public List<String> domains(){
    m001(); Map<String,Integer> map = domain(this.all);
    int res = max(map.values());
    return asList(String.valueOf(res));
  }// main
  public void test(){ List<String> res = domains(); System.out.println(res);}
}
// input product
class User extends UserFile{
  List<String> lines; final FromString fs = new FromString();
  public User() throws IOException{
    this.lines = new ArrayList<>();
    BufferedReader br = new BufferedReader(new InputStreamReader(System.in)); 
    String line = br.readLine(); lines.add(line.trim());
    line = br.readLine(); lines.add(line.trim());
    // int lim = lines.get(0).length() - 1;
    // for(int i= 0; i<lim; i++){ lines.add(br.readLine().trim());}
    br.close();
  }
  public List<String> list(){ return this.lines;}
}
